File: src/problem-2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Searching = void 0;
var TABLE_SIZE = 5;
var hashFunction = function (key) { return key.charCodeAt(0) % TABLE_SIZE; };
var createHashTable = function () {
    return new Map(Array.from({ length: TABLE_SIZE }, function (_, i) { return [i, []]; }));
};
var insert = function (table, key, value) {
    var _a;
    var index = hashFunction(key);
    (_a = table.get(index)) === null || _a === void 0 ? void 0 : _a.unshift([key, value]);
    return table;
};
var search = function (table, key) {
    var _a, _b, _c;
    var index = hashFunction(key);
    return (_c = (_b = (_a = table.get(index)) === null || _a === void 0 ? void 0 : _a.find(function (_a) {
        var storedKey = _a[0];
        return storedKey === key;
    })) === null || _b === void 0 ? void 0 : _b[1]) !== null && _c !== void 0 ? _c : null;
};
var display = function (table) {
    table.forEach(function (bucket, i) {
        console.log("Bucket ".concat(i, ":"), bucket.map(function (_a) {
            var k = _a[0], v = _a[1];
            return "[".concat(k, " -> ").concat(v, "]");
        }).join(" "));
    });
};
var Searching = function () {
    var table = createHashTable();
    var entries = [
        ["S", 0], ["E", 1], ["A", 2], ["R", 3], ["C", 4],
        ["H", 5], ["X", 7], ["M", 9], ["P", 10], ["L", 11]
    ];
    entries.forEach(function (_a) {
        var key = _a[0], value = _a[1];
        return (table = insert(table, key, value));
    });
    display(table);
    var searchKey = "C";
    var result = search(table, searchKey);
    console.log(result !== null ? "\nKey ".concat(searchKey, " found with value: ").concat(result) : "\nKey ".concat(searchKey, " not found."));
};
exports.Searching = Searching;

File: problem-2.cpp
#include <iostream>
#include <list>
#include <vector>

class HashTable {
private:
    static const int TABLE_SIZE = 5; // Number of buckets (small for demonstration)
    std::vector<std::list<std::pair<char, int>>> table;

    // Hash function
    int hashFunction(char key) {
        return key % TABLE_SIZE;  // Simple modulo hash function
    }

public:
    // Constructor
    HashTable() {
        table.resize(TABLE_SIZE);
    }

    // Insert key-value pair
    void insert(char key, int value) {
        int index = hashFunction(key);
        table[index].push_front({key, value});  // Insert at the front of the list
    }

    // Search for a value by key
    int search(char key) {
        int index = hashFunction(key);
        for (const auto &entry : table[index]) {
            if (entry.first == key) {
                return entry.second;  // Return value if key is found
            }
        }
        return -1;  // Key not found
    }

    // Display the hash table
    void display() {
        for (int i = 0; i < TABLE_SIZE; i++) {
            std::cout << "Bucket " << i << ": ";
            for (const auto &entry : table[i]) {
                std::cout << "[" << entry.first << " -> " << entry.second << "] ";
            }
            std::cout << "\n";
        }
    }
};

int main() {
    HashTable ht;

    // Inserting values from the image example
    ht.insert('S', 0);
    ht.insert('E', 1);
    ht.insert('A', 2);
    ht.insert('R', 3);
    ht.insert('C', 4);
    ht.insert('H', 5);
    ht.insert('X', 7);
    ht.insert('M', 9);
    ht.insert('P', 10);
    ht.insert('L', 11);

    // Displaying the hash table
    ht.display();

    // Searching for keys
    char searchKey = 'C';
    int result = ht.search(searchKey);
    if (result != -1)
        std::cout << "\nKey " << searchKey << " found with value: " << result << "\n";
    else
        std::cout << "\nKey " << searchKey << " not found.\n";

    return 0;
}
